### DISCUSSION TITLE
使用 processor_split_char 分隔日志字段的时候，如果最后一个字段为空，会被丢弃

### DISCUSSION THREAD

**Discussion by gohonsen** [alibaba/ilogtail#871] [Help] [answered]
采集 Java 多行日志后，使用 processor_split_char 处理分隔，当最后一个字段为空的时候，输出的 content 没有对应的字段，而中间字段为空，则没问题。
采集文件配置：
```
enable: true
inputs:
  - Type: file_log
    LogPath: /tmp
    FilePattern: test-log.ing
processors:
  - Type: processor_split_log_regex
    SplitRegex: \d+-\d+-\d+\s\d+:\d+:\d+\.\d+\|.*
    SplitKey: content
    PreserveOthers: true
  - Type: processor_split_char
    SourceKey: content
    SplitSep: "|"
    SplitKeys:
      - timestamp
      - level_trace
      - pid
      - method
      - class_line
      - thread
      - message
      - exception
flushers:
  - Type: flusher_stdout
    OnlyStdout: true
``` 
日志内容:
echo -e '2023-05-24 14:00:00.643| INFO [test,f3e1d93809be88e7,f3e1d93809be88e7,false]|2724|setDB|c.s.n.a.c.m.DataSourceContextHolder:18|[HTTP-exec-10]||Message没内容Exception有内容' >>  /tmp/test-log.ing
输出: 
```
{
    "__tag__:__path__": "/tmp/test-log.ing",
    "timestamp": "2023-05-24 14:00:00.643",
    "level_trace": " INFO [test,f3e1d93809be88e7,f3e1d93809be88e7,false]",
    "pid": "2724",
    "method": "setDB",
    "class_line": "c.s.n.a.c.m.DataSourceContextHolder:18",
    "thread": "[HTTP-exec-10]",
    "message": "",
    "exception": "Message没内容Exception有内容",
    "__time__": "1684908496"
}
```
echo -e '2023-05-24 14:00:00.643| INFO [test,f3e1d93809be88e7,f3e1d93809be88e7,false]|2724|setDB|c.s.n.a.c.m.DataSourceContextHolder:18|[HTTP-exec-10]|Message有内容Exception没内容|' >>  /tmp/test-log.ing
输出:
```
{
    "__tag__:__path__": "/tmp/test-log.ing",
    "timestamp": "2023-05-24 14:00:00.643",
    "level_trace": " INFO [test,f3e1d93809be88e7,f3e1d93809be88e7,false]",
    "pid": "2724",
    "method": "setDB",
    "class_line": "c.s.n.a.c.m.DataSourceContextHolder:18",
    "thread": "[HTTP-exec-10]",
    "message": "Message有内容Exception没内容",
    "__time__": "1684908584"
}
```
可以看到当 exception 字段为空时，输出的 json 里面直接丢掉了 exception 这个key ，而中间的message字段为空，key 却能保留。

![ilogtail-1](https://github.com/alibaba/ilogtail/assets/15814866/bf1fc731-4270-4955-85da-33d245adf567)

**Comment by linrunqi08** [comment, marked as answer]
目前的实现的确会有这个问题，可以看下代码https://github.com/alibaba/ilogtail/blob/main/plugins/processor/split/char/split_char.go#L134 ，如果分隔符正好在最后一个字符的话，最后一个key是不会处理的。
@gohonsen  是否有兴趣来修改一下呢，golang这部分修改与测试都比较容易上手
